pandas/core/algorithms.py

```diff
@@ -947,11 +947,26 @@ def select_n_frame(frame, columns, n, method, keep):
     from pandas.core.series import Series
     if not is_list_like(columns):
         columns = [columns]
-    columns = list(columns)
-    ser = getattr(frame[columns[0]], method)(n, keep=keep)
-    if isinstance(ser, Series):
-        ser = ser.to_frame()
-    return ser.merge(frame, on=columns[0], left_index=True)[frame.columns]
+    else:
+        columns = list(columns)
+
+    ascending = method == 'nsmallest'
+    tmp = Series(frame.index)
+    frame.reset_index(inplace=True, drop=True)
```

Comment:
I need to reorder the incoming frame based on the specified sorting method. I can't do that using the original index as it might contain duplicates which gives rise to the original issue. Therefore I save the original index to temp and reset the index of frame inplace so we're not needlessly copying the entire frame.